File: test_reddit_api.py
#!/usr/bin/env python3
"""
Reddit API Scope Testing Script
This script helps debug Reddit OAuth scope issues by testing API endpoints directly.
"""

import requests
import sys

def test_reddit_endpoint(access_token, endpoint, description):
    """Test a Reddit API endpoint with the given access token."""
    print(f"\n🔍 Testing: {description}")
    print(f"📡 Endpoint: {endpoint}")
    
    headers = {
        'Authorization': f'Bearer {access_token}',
        'User-Agent': 'web:com.skoop.app:v1.0.0 (by /u/skoop_support)'
    }
    
    try:
        response = requests.get(endpoint, headers=headers)
        print(f"📊 Status: {response.status_code}")
        
        if response.status_code == 200:
            data = response.json()
            print(f"✅ Success! Response keys: {list(data.keys())}")
            if 'data' in data and 'children' in data['data']:
                print(f"📈 Items found: {len(data['data']['children'])}")
        else:
            print(f"❌ Error: {response.status_code}")
            try:
                error_data = response.json()
                print(f"🔍 Error details: {error_data}")
            except:
                print(f"🔍 Error text: {response.text}")
                
    except Exception as e:
        print(f"💥 Exception: {e}")

def main():
    if len(sys.argv) != 2:
        print("Usage: python test_reddit_api.py <ACCESS_TOKEN>")
        print("\nGet your access token from the connected_accounts table in Supabase")
        sys.exit(1)
    
    access_token = sys.argv[1]
    
    print("🔴 Reddit API Scope Testing")
    print("="*50)
    
    # Test 1: User info (should work with 'identity' scope)
    test_reddit_endpoint(
        access_token,
        "https://oauth.reddit.com/api/v1/me",
        "User Info (identity scope)"
    )
    
    # Test 2: User overview (requires 'history' scope)
    test_reddit_endpoint(
        access_token,
        "https://oauth.reddit.com/user/me/overview",
        "User Overview (history scope)"
    )
    
    # Test 3: Saved items - various endpoints
    endpoints_to_test = [
        ("https://oauth.reddit.com/user/me/saved", "Saved Items - /user/me/saved"),
        ("https://oauth.reddit.com/user/me/saved.json", "Saved Items - /user/me/saved.json"),
        ("https://oauth.reddit.com/u/me/saved", "Saved Items - /u/me/saved"),
        ("https://oauth.reddit.com/u/me/saved.json", "Saved Items - /u/me/saved.json"),
    ]
    
    for endpoint, description in endpoints_to_test:
        test_reddit_endpoint(access_token, endpoint, description)
    
    print("\n" + "="*50)
    print("🎯 Diagnosis Guide:")
    print("- If user info works but saved items fail: Missing 'history' or 'save' scopes")
    print("- If all endpoints fail: Token expired or invalid")
    print("- If 403 errors: App doesn't have required permissions")
    print("- If 400 errors: Wrong endpoint or scope issues")

if __name__ == "__main__":
    main() 

File: test_reddit_api.js
#!/usr/bin/env node

/**
 * Manual Reddit API Tester
 * Usage: node test_reddit_api.js YOUR_ACCESS_TOKEN
 * 
 * This script tests various Reddit API endpoints to identify which one works
 * and provides detailed debugging information.
 */

const https = require('https');
const { URL } = require('url');

if (process.argv.length < 3) {
  console.error('❌ Usage: node test_reddit_api.js YOUR_ACCESS_TOKEN');
  process.exit(1);
}

const accessToken = process.argv[2];

const headers = {
  'Authorization': `Bearer ${accessToken}`,
  'User-Agent': 'web:com.skoop.app:v1.0.0 (by /u/skoop_support)',
  'Accept': 'application/json'
};

async function makeRequest(url) {
  return new Promise((resolve, reject) => {
    const parsedUrl = new URL(url);
    
    const options = {
      hostname: parsedUrl.hostname,
      port: parsedUrl.port || 443,
      path: parsedUrl.pathname + parsedUrl.search,
      method: 'GET',
      headers: headers
    };

    console.log(`🔍 Testing: ${url}`);
    console.log(`🔍 Headers: ${JSON.stringify({
      ...headers,
      'Authorization': 'Bearer [REDACTED]'
    }, null, 2)}`);

    const req = https.request(options, (res) => {
      let data = '';
      
      console.log(`📊 Status: ${res.statusCode}`);
      console.log(`📊 Response Headers: ${JSON.stringify(res.headers, null, 2)}`);
      
      res.on('data', (chunk) => {
        data += chunk;
      });
      
      res.on('end', () => {
        try {
          const jsonData = JSON.parse(data);
          resolve({
            status: res.statusCode,
            headers: res.headers,
            data: jsonData,
            success: res.statusCode >= 200 && res.statusCode < 300
          });
        } catch (error) {
          resolve({
            status: res.statusCode,
            headers: res.headers,
            data: data,
            success: res.statusCode >= 200 && res.statusCode < 300,
            parseError: error.message
          });
        }
      });
    });

    req.on('error', (error) => {
      reject(error);
    });

    req.end();
  });
}

async function testRedditAPI() {
  console.log('🚀 Starting Reddit API Test Suite\n');
  
  // Step 1: Verify token with user info
  console.log('=== STEP 1: Token Verification ===');
  try {
    const userResult = await makeRequest('https://oauth.reddit.com/api/v1/me');
    
    if (userResult.success) {
      console.log('✅ Token is valid!');
      console.log(`👤 Username: ${userResult.data.name}`);
      console.log(`🔢 User ID: ${userResult.data.id}`);
      console.log(`📧 Has Email: ${userResult.data.has_verified_email}`);
      console.log(`🎂 Created: ${new Date(userResult.data.created_utc * 1000).toISOString()}\n`);
    } else {
      console.error('❌ Token validation failed:', userResult.data);
      return;
    }
  } catch (error) {
    console.error('💥 Token validation error:', error.message);
    return;
  }
  
  // Step 2: Test saved items endpoints
  console.log('=== STEP 2: Saved Items Endpoints ===');
  
  const endpointsToTest = [
    'https://oauth.reddit.com/user/me/saved.json?raw_json=1&limit=5',
    'https://oauth.reddit.com/user/me/saved.json?limit=5',
    'https://oauth.reddit.com/user/me/saved?raw_json=1&limit=5',
    'https://oauth.reddit.com/user/me/saved?limit=5',
    'https://oauth.reddit.com/user/me/saved',
    'https://oauth.reddit.com/api/v1/me/saved?limit=5'
  ];
  
  let workingEndpoint = null;
  
  for (const endpoint of endpointsToTest) {
    console.log(`\n--- Testing: ${endpoint} ---`);
    
    try {
      const result = await makeRequest(endpoint);
      
      if (result.success) {
        console.log('✅ SUCCESS!');
        console.log(`📊 Items found: ${result.data?.data?.children?.length || 0}`);
        
        if (result.data?.data?.children?.length > 0) {
          const firstItem = result.data.data.children[0];
          console.log(`📝 First item type: ${firstItem.kind}`);
          console.log(`📝 First item title: ${firstItem.data.title || firstItem.data.body?.substring(0, 50) || 'N/A'}`);
        }
        
        workingEndpoint = endpoint;
        break;
      } else {
        console.log('❌ FAILED');
        console.log(`📊 Error: ${JSON.stringify(result.data, null, 2)}`);
      }
    } catch (error) {
      console.log('💥 Network error:', error.message);
    }
    
    // Small delay between requests
    await new Promise(resolve => setTimeout(resolve, 1000));
  }
  
  // Step 3: Summary
  console.log('\n=== SUMMARY ===');
  if (workingEndpoint) {
    console.log(`✅ Working endpoint found: ${workingEndpoint}`);
    console.log('🎯 Use this endpoint in your Edge Function!');
  } else {
    console.log('❌ No working endpoints found');
    console.log('🔍 This suggests an OAuth scope issue');
    console.log('📋 Required scopes: identity, history, save, read');
    console.log('💡 Check your Reddit app configuration at: https://www.reddit.com/prefs/apps');
  }
}

testRedditAPI().catch(console.error); 